Source/diffusion/Castro_diffusion.cpp

```diff
@@ -13,54 +13,57 @@ Castro::construct_old_diff_source(Real time, Real dt)
 {
     int ng = Sborder.nGrow();
 
-    old_sources[diff_src]->setVal(0.0);    
-
     MultiFab TempDiffTerm(grids, dmap, 1, 1);
     MultiFab SpecDiffTerm(grids, dmap, NumSpec, 1);
     MultiFab ViscousTermforMomentum(grids, dmap, BL_SPACEDIM, 1);
     MultiFab ViscousTermforEnergy(grids, dmap, 1, 1);
 
-    add_temp_diffusion_to_source(*old_sources[diff_src], TempDiffTerm, time, 1);
+    add_temp_diffusion_to_source(old_sources, TempDiffTerm, time, 1);
 
 #if (BL_SPACEDIM == 1)
-    add_spec_diffusion_to_source(*old_sources[diff_src], SpecDiffTerm, time, 1);
-    add_viscous_term_to_source(*old_sources[diff_src], ViscousTermforMomentum, ViscousTermforEnergy, time);
+    add_spec_diffusion_to_source(old_sources, SpecDiffTerm, time, 1);
+    add_viscous_term_to_source(old_sources, ViscousTermforMomentum, ViscousTermforEnergy, time);
 #endif
-
-    old_sources[diff_src]->FillBoundary(geom.periodicity());
 }
 
 void
 Castro::construct_new_diff_source(Real time, Real dt)
 {
     int ng = 0;
 
-    new_sources[diff_src]->setVal(0.0);
-
     MultiFab TempDiffTerm(grids, dmap, 1, 1);
     MultiFab SpecDiffTerm(grids, dmap, NumSpec, 1);
     MultiFab ViscousTermforMomentum(grids, dmap, BL_SPACEDIM, 1);
     MultiFab ViscousTermforEnergy(grids, dmap, 1, 1);
 
-    add_temp_diffusion_to_source(*new_sources[diff_src], TempDiffTerm, time, 0);
+    Real mult_factor = 0.5;
+
+    add_temp_diffusion_to_source(new_sources, TempDiffTerm, time, 0, mult_factor);
 
 #if (BL_SPACEDIM == 1)
-    add_spec_diffusion_to_source(*new_sources[diff_src], SpecDiffTerm, time, 0);
-    add_viscous_term_to_source(*new_sources[diff_src], ViscousTermforMomentum, ViscousTermforEnergy, time);
+    add_spec_diffusion_to_source(new_sources, SpecDiffTerm, time, 0, mult_factor);
+    add_viscous_term_to_source(new_sources, ViscousTermforMomentum, ViscousTermforEnergy, time, mult_factor);
 #endif
 
     // Time center the source term.
 
-    new_sources[diff_src]->mult(0.5);
+    mult_factor = -0.5;
+    Real old_time = time - dt;
+
```

Comment:
how does this work -- we are no longer multiplying by +1/2 -- is that absorbed in the Fortran now?

Reply:
The mult_factor is used in the Saxpy that is done at the end of add_*_diffusion_to_source; so the net result is that we get 0.5 * new_time - 0.5 * old_time, which is the same as we had before.